lib/lotus/application_name.rb

```diff
@@ -0,0 +1,95 @@
+module Lotus
+  # An application name.
+  #
+  # @since x.x.x
+  class ApplicationName
+
+    # A list of words that are prohibited from forming the application name
+    #
+    # @since x.x.x
+    INVALID_WORDS = %w(lotus).freeze
+
+
+    # Initialize and check against reserved words
+    #
+    # An application name needs to be translated in quite a few ways:
+    # First, it must be checked against a list of reserved words and rejected
+    # if it is invalid. Secondly, assuming it is not invalid, it must be able
+    # to be output roughly as given, but with the following changes:
+    #
+    # 1. downcased,
+    # 2. with surrounding spaces removed,
+    # 3. with internal whitespace rendered as underscores
+    # 4. with underscores de-duplicated
+    #
+    # which is the default output:
+    #
+    #     ApplicationName.new("my-App ").to_s # => "my-app"
+    #
+    # It must also be transformable into an environment variable:
+    #
+    #     ApplicationName.new("my-app").to_env_s => "MY_APP"
+    #
+    # @return [Lotus::ApplicationName] a new instance of the application name
+    #
+    # @since x.x.x
+    def initialize(name)
+      @name = sanitize(name)
+      ensure_validity!
+    end
+
+
+    # Returns the cleaned application name.
+    #
+    # @since x.x.x
+    def to_s
+      @name
+    end
+
+
+    # Returns the application name uppercased with non-alphanumeric characters
+    # as underscores.
+    #
+    # @since x.x.x
+    def to_env_s
+      @name.upcase.gsub(/\W/, '_')
+    end
+
+
+    # Returns true if a potential application name matches one of the reserved
+    # words.
+    #
+    # @since x.x.x
+    def self.invalid?(name)
+      INVALID_WORDS.include?(name)
+    end
+
+
+    private
+
+    # Raises RuntimeError with explanation if the provided name is invalid.
+    #
+    # @api private
+    # @since x.x.x
+    def ensure_validity!
+      if self.class.invalid?(@name)
+        raise RuntimeError,
+          "application name must not be any one of the following: " +
+          INVALID_WORDS.join(", ")
+      end
+    end
+
+
+    # Cleans a string to be a functioning application name.
+    #
+    # @api private
+    # @since x.x.x
+    def sanitize(name)
+      name
```

Comment:
Doesn't quite do the same thing:
```
1) Failure:
Lotus::ApplicationName::formats::#to_s#test_0003_renders internal spaces as underscores [/home/rob/Sites/l/lotus/lotus/test/application_name_test.rb:21]:
Expected: "my_app"
  Actual: "my app"


  2) Failure:
Lotus::ApplicationName::formats::#to_s#test_0004_renders with underscores de-duplicated [/home/rob/Sites/l/lotus/lotus/test/application_name_test.rb:26]:
Expected: "my_app"
  Actual: "my _app"


  3) Failure:
Lotus::ApplicationName::formats::#to_s#test_0002_renders trimmed [/home/rob/Sites/l/lotus/lotus/test/application_name_test.rb:16]:
Expected: "my-app"
  Actual: " my-app "
```

The underscoring is part of it, yes, but a big part of it is also sanitization of potential incorrect input. Not opposed to using a `Lotus::Util::String` method or creating one, though.